mirrord/agent/src/steal/connections/filtered.rs

```diff
@@ -0,0 +1,1524 @@
+
+    type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send>>;
+
+    fn call(&self, request: Request<Incoming>) -> Self::Future {
+        let service = self.clone();
+
+        Box::pin(async move { service.handle_request(request).await })
+    }
+}
+
+/// Manages a filtered stolen connection.
+///
+/// Uses a custom [`Service`] implementation to handle HTTP protocol details, change the raw IO
+/// stream into a series requests and provide responses.
+pub struct FilteredStealTask<T> {
+    connection_id: ConnectionId,
+    /// Original destination of the stolen connection. Used when passing through HTTP requests that
+    /// don't not match any filter in [`Self::filters`].
+    original_destination: SocketAddr,
+
+    /// Stealer client to [`HttpFilter`] mapping. Allows for routing HTTP requests to correct
+    /// stealer clients.
+    ///
+    /// # Note
+    ///
+    /// This mapping is shared via [`Arc`], allowing for dynamic updates from the outside.
+    /// This allows for *injecting* new stealer clients into exisiting connections.
+    filters: Arc<DashMap<ClientId, HttpFilter>>,
+
+    /// Stealer client to subscription state mapping.
+    /// 1. `true` -> client is subscribed
+    /// 2. `false` -> client has unsubscribed or we sent [`ConnectionMessageOut::Closed`].
+    /// 3. `None` -> client does not know about this connection at all
+    subscribed: HashMap<ClientId, bool>,
+
+    /// For receiving [`Request`]s extracted by the [`FilteringService`].
+    requests_rx: Receiver<ExtractedRequest>,
+
+    /// 1. Handle for the [`tokio::task`] that polls the [`hyper`] connection of related
+    ///    [`FilteringService`].
+    /// 2. [`DropGuard`] for this [`tokio::task`], so that it aborts when this struct is dropped.
+    hyper_conn_task: Option<(JoinHandle<Option<UpgradedConnection>>, DropGuard)>,
+
+    /// Requests blocked on stealer clients' responses.
+    blocked_requests: HashMap<(ClientId, RequestId), oneshot::Sender<RequestHandling>>,
+
+    /// Id of the next HTTP request that will be intercepted.
+    next_request_id: RequestId,
+
+    /// For safely downcasting the IO stream after an HTTP upgrade. See [`Upgraded::downcast`].
+    _io_type: PhantomData<fn() -> T>,
+}
+
+impl<T> FilteredStealTask<T>
+where
+    T: AsyncRead + AsyncWrite + Unpin + Send + 'static,
+{
+    /// Creates a new instance of this task. The task will manage the connection given as `io` and
+    /// use the provided `filters` for matching incoming [`Request`]s with stealing clients.
+    ///
+    /// The task will not run yet, see [`Self::run`].
+    pub fn new(
+        connection_id: ConnectionId,
+        filters: Arc<DashMap<ClientId, HttpFilter>>,
+        original_destination: SocketAddr,
+        http_version: HttpVersion,
+        io: T,
+    ) -> Self {
+        let (upgrade_tx, mut upgrade_rx) = mpsc::channel(1);
+        let (requests_tx, requests_rx) = mpsc::channel(8);
```

Comment:
Is there a reason for this specific channel size?

Reply:
Dunno, more of a guess. Limiting upgrade channel to 1 makes sense, because it could be a `oneshot` anyway (I only made it an `mpsc` to make the code less ugly, using a `oneshot` required `Arc<Mutex<Option<...>>>` wrapper. Limiting requests channel to 8 gives as at most 8 concurrent requests that wait to be handled by users' apps. Do you think it's too low?

Reply:
It might be a little small though, I would Increase it to probably around 100 and set it as a const so it will be understood that it's a "concurrency" limit